service/matching/taskListManager.go

```diff
@@ -42,20 +42,95 @@ const (
 	done time.Duration = -1
 )
 
+// NOTE: Is this good enough for stress tests?
+const (
+	_maxDispatchDefault = 100000.0
+	_dispatchLimitTTL   = time.Second
+)
+
 type taskListManager interface {
 	Start() error
 	Stop()
 	AddTask(execution *s.WorkflowExecution, taskInfo *persistence.TaskInfo) error
 	GetTaskContext(ctx context.Context) (*taskContext, error)
 	SyncMatchQueryTask(ctx context.Context, queryTask *queryTaskInfo) error
 	CancelPoller(pollerID string)
+	UpdateMaxDispatch(maxDispatchPerSecond *float64)
 	String() string
 }
 
-func newTaskListManager(e *matchingEngineImpl, taskList *taskListID, config *Config) taskListManager {
+type rateLimiter struct {
+	maxDispatchPerSecond *float64
+	tokenBucket          common.TokenBucket
+	ttl                  *time.Timer
```

Comment:
consider just using epoch time to check for ttl expiration instead of using a timer i.e. just store expiration as now().nanos+ttl and within your UpdateMaxDispatch just check if time.Now() > expiration. 

Timers use the global heap within the go runtime (lock protected afaik) i.e. more expensive than time.Now()

Reply:
I would keep it simple here, not sure the perf is significant enough to make it more complicated. Also timer uses channels internally, no lock as far as I can see https://golang.org/src/time/sleep.go

Reply:
madhu - storing the expiry as epoch is a well understood pattern on the server side. And given that this is just couple of lines of code change and it buys us efficiency, why wait until there is a perf issue, lets just do it and be done with it.

Reply:
Timer is the recommended pattern and simple/easy to understand. What efficiency does epoch buy us? What is the perf advantage here exactly?

Reply:
recommended ? see this https://github.com/golang/go/issues/15133